src/file.cr

```diff
@@ -672,6 +672,80 @@ class File < IO::FileDescriptor
     end
   end
 
+  # Ensures the content written to the file descriptor is written completely or not at all
+  # preventing corruption of the file.
+  #
+  # If a file is being created, its initial permissions may be set using the *perm* parameter.
+  # Then the given block will be passed the opened file descriptor as an argument, the file will
+  # be automatically closed and saved when the block returns.
+  #
+  # This is done by saving the new contents at temporary path. When the new content is
+  # successfully written the temporary path is changed to the provided path ensuring the data is
+  # not corrupted. If the write fails the temporary file is deleted.
+  def self.atomic_write(path : String, perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil, *, append : Bool = false, &block : IO::FileDescriptor -> Nil) : Bool
+    atomic_path = new_atomic_path(path)
+    success_flag = false
+
+    begin
+      open(atomic_path, "w", perm, encoding, invalid) do |fd|
+        fd.flock_exclusive() do
+          open(path, "r") { |src| IO.copy(src, fd) } if append
+          yield(fd)
+          fd.flush
+          success_flag = true
+        end
+      end
+    ensure
+      if (success_flag)
+        atomic_install(atomic_path, path)
+      else
+        delete(atomic_path)
+      end
+    end
+
+    return success_flag
+  end
+
+  # Writes the provided content completely or not at all preventing file corruption.
+  #
+  # This is preformed in the same way as `atomic_write` with block.
+  #
+  # NOTE: If the content is a `Slice(UInt8)`, those bytes will be written.
+  # If it's an `IO`, all bytes from the `IO` will be written.
+  # Otherwise, the string representation of *content* will be written
+  # (the result of invoking `to_s` on *content*).
+  def self.atomic_write(path : String, content, perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil) : Bool
+    atomic_write(path, perm, encoding, invalid) do |fd|
+      case content
+      when Bytes then fd.write(content)
+      when IO    then IO.copy(content, fd)
```

Comment:
Consider making different method overloads

Reply:
I just copied `File.write()` which uses this method. I suspect its for verbosity reasons.